feat: add support for `getLocFromIndex` and `getIndexFromLoc`

lumirlumir opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [OpenJS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Update or create tests
    - If performance-related, include a benchmark
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What is the purpose of this pull request?

Hello,

In this PR, I've completed implemention of `getLocFromIndex()` and `getIndexFromLoc()`.

I based the implementation on the current [JavaScript `SourceCode` class](https://github.com/eslint/eslint/blob/main/lib/languages/js/source-code/source-code.js#L738-L846) and incorporated the fast lookup algorithm discussed in [eslint/eslint#19782](https://github.com/eslint/eslint/pull/19782).

The logic of `getLocFromIndex()` and `getIndexFromLoc()` is nearly identical to that in the JavaScript `SourceCode` class, with the key difference being that it accounts for the `lineStart` and `columnStart` values initialized in the constructor.

---

##### `lineStart` and `columnStart`

`TextSourceCodeBase` should accept `lineStart` and `columnStart`, as these can vary depending on the language, and both `getLocFromIndex()` and `getIndexFromLoc()` rely on them for accurate calculations. Here are some examples:

- https://github.com/eslint/css/blob/main/src/languages/css-language.js#L69-L79
- https://github.com/eslint/markdown/blob/main/src/language/markdown-language.js#L91-L101
- https://github.com/eslint/json/blob/main/src/languages/json-language.js#L44-L54
- https://github.com/yeonjuan/html-eslint/blob/main/packages/eslint-plugin/lib/languages/html-language.js#L21-L31

#### What changes did you make? (Give an overview)

I've completed implemention of `getLocFromIndex()` and `getIndexFromLoc()`.

#### Related Issues

fixes: https://github.com/eslint/rewrite/issues/166
refs: https://github.com/eslint/markdown/pull/376, https://github.com/eslint/css/pull/167, https://github.com/eslint/json/pull/109, https://github.com/eslint/eslint/pull/19782

<!-- include tags like "fixes #123" or "refs #123" -->

#### Is there anything you'd like reviewers to focus on?

<!-- markdownlint-disable-file MD004 -->

#### Prerequisites

- [x] https://github.com/eslint/eslint/pull/19782

nzakas commented on packages/plugin-kit/src/source-code.js:
```diff
@@ -243,12 +287,30 @@ export class TextSourceCodeBase {
 		this.ast = ast;
 		this.text = text;
-		this.#lines = text.split(lineEndingPattern);
+		this.#lineStart = lineStart;
+		this.#columnStart = columnStart;
```
I don't think we need to pass these in. We should be able to get this information off of `ast`.

nzakas reviewed (commented):
Thanks for taking a look at this. I left some comments throughout.

Overall, I'd like to think more about how to lazily generate the data needed. For instance, if someone calls `getLocFromIndex(5)`, we should only generate the line endings up to index 5. If someone then calls `getLocFromIndex(22)`, we can go on to generate line endings up to index 22.

That way, we're saving calculations until we really need them.

lumirlumir reviewed (commented):
It took a while, but I’ve finally finished implementing these two methods 😄

I’ve carefully reviewed the lazy calculation and addressed all the comments.

There might still be room for improvement—if so, I’d be happy to take a look.

(I didn’t go with the caching approach, as it turned out to be slightly slower in https://github.com/eslint/eslint/pull/19782)

lumirlumir commented:
@nzakas 

I've addressed all the comments and done a bit of refactoring to make the code clearer. 😄

lumirlumir commented on packages/plugin-kit/src/source-code.js:
```diff
@@ -281,25 +281,55 @@ export class TextSourceCodeBase {
+	 * @param {string} text The source text to parse into lines.
+	 * @returns {boolean} `true` if the text was successfully parsed into lines, `false` otherwise.
+	 */
+	#parseText(text) {
```
I've added a new commit https://github.com/eslint/rewrite/pull/212/commits/13014b45cfd5341eff789343d2d5d6fbbbca3ab1

lumirlumir commented on packages/plugin-kit/src/source-code.js:
```diff
@@ -348,22 +371,11 @@ export class TextSourceCodeBase {
-			Boolean(additionalLinesNeeded--) &&
-			(match = lineEndingPattern.exec(text))
+			this.#parseText(this.text.slice(this.#lineStartIndices.at(-1))) &&
+			Boolean(additionalLinesNeeded--)
```
I've added a new commit https://github.com/eslint/rewrite/pull/212/commits/3d3448ac5a60f982448a78abb3712cb69081aca9

mdjermanovic commented on packages/plugin-kit/README.md:
```diff
@@ -205,6 +205,8 @@ The `TextSourceCodeBase` class is intended to be a base class that has several o
 
 - `lines` - an array of text lines that is created automatically when the constructor is called.
 - `getLoc(node)` - gets the location of a node. Works for nodes that have the ESLint-style `loc` property and nodes that have the Unist-style [`position` property](https://github.com/syntax-tree/unist?tab=readme-ov-file#position). If you're using an AST with a different location format, you'll still need to implement this method yourself.
+- `getLocFromIndex(index)` - Converts a source text index into a `{ line: number, column: number }` pair.
+- `getIndexFromLoc(loc)` - Converts a `{ line: number, column: number }` pair into a source text index.
```
Perhaps we should add a note that, in order for these to work, the root node must always cover the entire source code text?

mdjermanovic commented on packages/plugin-kit/src/source-code.js:
```diff
@@ -243,12 +281,108 @@ export class TextSourceCodeBase {
 		this.ast = ast;
 		this.text = text;
-		this.#lines = text.split(lineEndingPattern);
+		// Remove the global(`g`) flag from the `lineEndingPattern` to avoid issues with lastIndex.
+		this.#lineEndingPattern = new RegExp(
+			lineEndingPattern.source,
+			lineEndingPattern.flags.replace("g", ""),
+		);
```
The `y` flag also modifies `lastIndex`, so perhaps we should remove that flag too.

lumirlumir commented:
@mdjermanovic 

Thank you for your thorough review!

I've resolved all the comments and added the test cases you suggested in https://github.com/eslint/rewrite/pull/212#discussion_r2242088769.

The bug you pointed out in https://github.com/eslint/rewrite/pull/212#discussion_r2242088769 was caused by the following code:

```js
this.text.slice(this.#lineStartIndices.at(-1), index + 1)
```

Since I used the `slice` method with `index + 1`, it led to the error you mentioned.

I've now fixed this in https://github.com/eslint/rewrite/pull/212/commits/842dfb8631e9e594744f18752686f354d1f1f48f, so the error no longer occurs.

The other private lazy methods already use the approach I implemented in https://github.com/eslint/rewrite/pull/212/commits/842dfb8631e9e594744f18752686f354d1f1f48f, so the bug was only present in the `#ensureLineStartIndicesFromIndex(index)` method.
